source/Octopus.Tentacle/Configuration/Instances/ApplicationInstanceSelector.cs

```diff
@@ -64,30 +66,50 @@ ApplicationInstanceConfiguration LoadCurrentInstance()
         ApplicationInstanceConfiguration LoadInstance()
         {
             var appInstance = LocateApplicationPrimaryConfiguration();
-            EnsureConfigurationExists(appInstance.instanceName, appInstance.configurationpath);
 
-            log.Verbose($"Loading configuration from {appInstance.configurationpath}");
-            var writableConfig = new XmlFileKeyValueStore(fileSystem, appInstance.configurationpath);
-            
-            var aggregatedKeyValueStore = ContributeAdditionalConfiguration(writableConfig);
+            var (aggregatedKeyValueStore, writableConfig) = LoadConfigurationStore(appInstance);
+
             return new ApplicationInstanceConfiguration(appInstance.instanceName, appInstance.configurationpath, aggregatedKeyValueStore, writableConfig);
         }
 
-        void EnsureConfigurationExists(string? instanceName, string configurationPath)
+        (IKeyValueStore, IWritableKeyValueStore) LoadConfigurationStore((string? instanceName, string? configurationpath) appInstance)
         {
-            if (!fileSystem.FileExists(configurationPath))
+#if !NETFRAMEWORK
+            if (appInstance is { instanceName: not null, configurationpath: null } &&
+                KubernetesConfig.Namespace is {} @namespace)
             {
-                var message = !string.IsNullOrEmpty(instanceName)
-                    ? $"The configuration file for instance {instanceName} could not be located at the specified location {configurationPath}. " +
-                    "The file might have been manually removed without properly removing the instance and as such it is still listed as present." +
-                    "The instance must be created again before you can interact with it."
-                    : $"The configuration file at {configurationPath} could not be located at the specified location.";
-                    
-                throw new ControlledFailureException(message);
+                log.Verbose($"Loading configuration from ConfigMap for namespace {@namespace}");
+                var configMapWritableStore = new ConfigMapKeyValueStore(@namespace, log);
+                return (ContributeAdditionalConfiguration(configMapWritableStore), configMapWritableStore);
             }
+#endif
+            if (!ConfigurationFileExists(appInstance.configurationpath))
+            {
+                ThrowConfigurationMissingException(appInstance.instanceName, appInstance.configurationpath);
+            }
+
+            log.Verbose($"Loading configuration from {appInstance.configurationpath}");
+            var writable = new XmlFileKeyValueStore(fileSystem, appInstance.configurationpath!);
+            return (ContributeAdditionalConfiguration(writable), writable);
         }
```

Comment:
This i'm not so sure about, especially as we'll start to need DI injection 🤔 
Maybe we can use the factory pattern to get both DI & non-di here